tests/test_keras_integration.py

```diff
@@ -0,0 +1,140 @@
+    return doc
+
+
+def assert_good_external_format(expl, overlay):
+    """
+    Check properties of the formatted heatmap over the original image,
+    using external properties of the image,
+    such as dimensions, mode, type.
+    """
+    original = expl.image
+    # check external properties
+    assert isinstance(overlay, Image.Image)
+    assert overlay.width == original.width
+    assert overlay.height == original.height
+    assert overlay.mode == 'RGBA'
+
+
+def assert_attention_over_area(expl, area):
+    """
+    Check that the explanation 'expl' lights up the most over 'area', 
+    a tuple of (x1, x2, y1, y2), starting and ending points of the bounding rectangle
+    in the original image.
+    We make two assumptions in this test:
+    1. The model can classify the example image correctly.
+    2. The area specified by the tester over the example image covers the predicted class correctly.
+    """
+    image = expl.image
+    heatmap = expl.targets[0].heatmap
+
+    # fit heatmap over image
+    heatmap = expand_heatmap(heatmap, image, interpolation=Image.LANCZOS)
+    heatmap = np.array(heatmap)
+
+    # get a slice of the area
+    x1, x2, y1, y2 = area
+    crop = heatmap[y1:y2, x1:x2] # row-first ordering
+    # TODO: instead of hard-coding the height and width offsets
+    # it might be a better idea to use percentages
+    # this makes the tests independent of any resizing done on the image
+    # and the heatmap doesn't have to be resized
+    # however, it might be harder for the user to determine percentages
+
+    # check intensity
+    total_intensity = np.sum(heatmap)
+    crop_intensity = np.sum(crop)
+    p = total_intensity / 100 # -> 1% of total_intensity
+    crop_p = crop_intensity / p # -> intensity %
+    # at least 50% (need to experiment with this number)
+    assert 50 < crop_p
+
+    # Alternatively, check that the intensity over area 
+    # is greater than all other intensity:
+    # remaining_intensity = total_intensity - intensity
+    # assert remaining_intensity < total_intensity
+
+
+# area = (x1, x2, y1, y2)
+@pytest.mark.parametrize('area, targets', [
+    ((54, 170, 2, 100), None), # focus on the dog (pick top prediction)
+    ((44, 180, 130, 212), [imagenet_cat_idx]), # focus on the cat (supply prediction)
+])
+def test_image_classification(keras_clf, cat_dog_image, area, targets):
+    # check explanation
+    res = eli5.explain_prediction(keras_clf, cat_dog_image, targets=targets)
+    assert_attention_over_area(res, area)
+    
+    # check formatting
+    overlay = format_as_image(res)
+    assert_good_external_format(res, overlay)
+
+    # check show function
+    show_overlay = eli5.show_prediction(keras_clf, cat_dog_image, targets=targets)
+    assert_pixel_by_pixel_equal(overlay, show_overlay)
+
+
+@pytest.fixture(scope="function")
+def show_nodeps():
+    # set up
+    old_format_as_image = eli5.ipython.format_as_image
+    eli5.ipython.format_as_image = ImportError('mock test')
```

Comment:
that's an interesting trick :) I wonder if yield should be wrapped in try/finally, to make sure teardown code is called

Reply:
https://github.com/pytest-dev/pytest/issues/2508 and https://docs.pytest.org/en/latest/fixture.html#fixture-finalization-executing-teardown-code suggest the "addfinalizer method" to ensure tear-down executes even if:
* the test code fails
* the set up fails (currently tear down is not executed if this happens)

https://github.com/TeamHG-Memex/eli5/pull/315/commits/50f7000c2d78beaf491f1ab516e83277e2ddb5c6